### ISSUE TITLE
Mixing lru_cache and overload messed up typing inference

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#2414]
**Describe the bug**
If a function is declared with `@lru_cache` from `functools` and has `overload`, the overload is not respected.

**To Reproduce**
```python
from functools import lru_cache
from typing import overload, Literal, Union


@overload
def myfun(ab: Literal["a"]) -> str:
    ...


@overload
def myfun(ab: Literal["b"]) -> int:
    ...


@lru_cache(maxsize=20)
def myfun(ab: Literal["a", "b"]) -> Union[str, int]:
    if ab == "a":
        return "a string"
    elif ab == "b":
        return 35
    else:
        raise ValueError("bad argument " + ab)


astr: str = myfun("a")
anint: int = myfun("b")
```
reports
```
Expression of type "str | int" cannot be assigned to declared type "str"
  Type "str | int" cannot be assigned to type "str"
    "int" is incompatible with "str"
Expression of type "str | int" cannot be assigned to declared type "int"
  Type "str | int" cannot be assigned to type "int"
    "str" is incompatible with "int"
```
If you comment out the line with `@lru_cache`, then the errors go away

**Expected behavior**
No error from the type checker.

**VS Code extension or command-line**
VS Code pylance V2021.10.0

**Comment by erictraut** [comment]
I don't think an overload makes sense to use with a decorator. Overloads provide alternative signatures for a function, and a decorator transforms a function's signature to something completely different.

In this example, the decorator effectively overwrites the original `myfun` function with an object that has a `__call__` method.

It's the equivalent of this:

```python
def myfun_undecorated(ab: Literal["a", "b"]) -> Union[str, int]:
    ...
myfun = lru_cache(maxsize=20)(myfun_undecorated)
```

The type of `myfun` is `_lru_cache_wrapper[str | int]`. The overload is therefore lost when the decorator is applied.

I was curious what mypy does in this case, and it simply crashes (generates an internal error).

I wish I could offer a good workaround, but I don't see any in this case. Overloads and decorators don't work well together.

**Comment by jakebailey** [comment]
The workaround I can imagine is to do something like this instead:

```py
@overload
def myfun(ab: Literal["a"]) -> str:
    ...


@overload
def myfun(ab: Literal["b"]) -> int:
    ...

def myfun(*args, **kwargs):
    return _myfun(*args, **kwargs)

@lru_cache(maxsize=20)
def _myfun(ab: Literal["a", "b"]) -> Union[str, int]:
    if ab == "a":
        return "a string"
    elif ab == "b":
        return 35
    else:
        raise ValueError("bad argument " + ab)
```

Keeping the overloads and decorators separate to deal with the incompatibility of `overload` with other decorators.

**Comment by erictraut** [comment]
Ah, that's clever. Good suggestion.

**Comment by jakebailey** [comment]
(FWIW, the mypy issue is https://github.com/python/mypy/issues/9112.)

**Comment by Kache** [comment]
Happened upon this myself for `@ft.lru_cache`. However, I switched implementation to using [cachetools](https://pypi.org/project/cachetools/) and discovered it doesn't have this issue:

```py
import cachetools as ct

@overload
def variant_value(actor_id: str, /) -> str | None: ...
@overload
def variant_value(actor_id: str, default: str) -> str: ...

@ct.cached(ct.TTLCache(maxsize=8, ttl=10))
def variant_value(actor_id: str, default: str | None = None):
    return "dummy value"

foo = variant_value('adsf')  # str | None
bar = variant_value('adsf', 'blah') # str
```

Does that mean this issue can be fixed in typeshed's annotations for `@ft.lru_cache`?

**Comment by erictraut** [comment]
The `cachetools.cached` decorator doesn't modify the signature of the decorated function, whereas `functools.lru_cache` does. 

Yes, typeshed could change its definition of `functools.lru_cache` so it likewise doesn't modify the signature. But this would less accurately reflect the runtime implementation of `lru_cache`, so it would create other problems (false positive errors). You could ask the typeshed maintainers what they think.
